sphinx_immaterial/theme_result.py

```diff
@@ -0,0 +1,53 @@
+"""A directive designed to reduce example snippets duplication."""
+from docutils.parsers.rst import directives
+from docutils import nodes
+from sphinx.util.docutils import SphinxDirective
+from sphinx.directives.code import container_wrapper
+
+
+class ResultsDirective(SphinxDirective):
+    """Takes content as a rst code snippet and renders it after outputting it as a
+    literal code-block."""
+
+    has_content = True
+    optional_arguments = 1
+    final_argument_whitespace = True
+    option_spec = {
+        "class": directives.class_option,
+        "name": directives.unchanged,
+        "output-prefix": directives.unchanged,
+    }
+
+    def run(self):
+        """Run the directive."""
+        container_node = nodes.container(
+            "", classes=self.options.get("class", []) + ["results"]
+        )
+        code = "\n".join(self.content)
+        literal_node = nodes.literal_block(code, code)
+        literal_node["language"] = "rst"
+        if self.arguments:
+            literal_node = container_wrapper(self, literal_node, self.arguments[0])
+        container_node += literal_node
+
+        if "output-prefix" in self.options:
+            out_prefix = nodes.container("", classes=["results-prefix"])
+            prefix = self.options.get("output-prefix", "")
+            if not prefix:
+                prefix = "Which renders the following content:"
+            textnodes, _ = self.state.inline_text(prefix, self.lineno)
+            out_prefix += textnodes
+            container_node += out_prefix
+
+        # the `result` CSS class is theme specific (as used by mkdocs-material theme)
+        results_div = nodes.container("", classes=["result"])
```

Comment:
Why are custom CSS rules needed?

Reply:
Several reasons. You'll notice that the directive has a `output-prefix` option and takes an optional arg (which gets used as the snippet's caption). This diverges from upstream's usage of example snippets, and builds on the fact that we can do better with this dedicated directive. Not to mention, the docutils output structure is different from mkdocs output.

Reply:
Previously we just used existing directives and rST syntax to show examples, e.g.:


```rst
.. details:: Open by default
   :class: example
   :open:

   .. code-block:: rst

      .. details:: Open by default
         :class: example
         :open:
```

This didn't require any custom CSS.  I imagined that this `rst-example` directive would basically just serve as a shorthand to produce exactly the same docutils nodes as before, just without the duplication.

I see that you have added some additional features, like `output-prefix`, but it seems like that could also be reasonably formatted just using existing nodes.

In general it just seems like a good idea to try to be minimal in the use of additional CSS and custom nodes if we can accomplish the same thing just by composing existing directives, as that keeps the theme simpler and there is less to maintain and fewer interactions between things to worry about.

Reply:
If you look at the mkdocs-material docs, you'll notice there is a border around the example snippets' output. I wanted to mimic that style (which was apparently inherited). 

However, the fact that literal blocks don't automatically have a `highlight` class (rather it gets a `highlight-xxx` class) caused the upstream CSS to not get used.

Reply:
Oh I see.  I looked at the mkdocs-material docs source code and I see that there is not in fact any special markdown support for the `results` class -- the mkdocs-material docs just use literal HTML to specify that.

In that case, rather than duplicate the styles, perhaps just modify the existing styles in pymdownx/_highlight.scss so that they work for your extension.

Reply:
That's reasonable. Again, I was trying to avoid merge conflicts there.

Reply:
Yeah makes sense, but if the upstream CSS changes it is actually better to have a merge conflict so that we know about it and make sure our modifications work with it.

Reply:
I'm resolving this as I've moved the custom CSS to _highlight.scss. Please review changes there, and let me know if there are any further requests about the custom CSS regarding `.results`, `.results` or `.results-prefix` (directive-specific div containers), etc.